content/style/go.md

```diff
@@ -1,3 +1,1659 @@
+// LogToStdout=0, LogToFile=1, LogToRemote=2
+```
+
+### Error Types
+
+There are various options for declaring errors:
+
+* [`errors.New`] for errors with simple static strings
+* [`fmt.Errorf`] for formatted error strings
+* Custom types that implement an `Error()` method
+* Wrapped errors using [`"pkg/errors".Wrap`]
+
+When returning errors, consider the following to determine the best choice:
+
+* Is this a simple error that needs no extra information? If so, [`errors.New`] should suffice.
+* Do the clients need to detect and handle this error? If so, you should use a custom type, and implement the `Error()` method.
+* Are you propagating an error returned by a downstream function? If so, check the [section on error wrapping](#error-wrapping).
+* Otherwise, [`fmt.Errorf`] is okay.
+
+  [`errors.New`]: https://golang.org/pkg/errors/#New
+  [`fmt.Errorf`]: https://golang.org/pkg/fmt/#Errorf
+  [`"pkg/errors".Wrap`]: https://godoc.org/github.com/pkg/errors#Wrap
+
+If the client needs to detect the error, and you have created a simple error using [`errors.New`], use a var for the error.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+// package foo
+
+func Open() error {
+  return errors.New("could not open")
+}
+
+// package bar
+
+func use() {
+  if err := foo.Open(); err != nil {
+    if err.Error() == "could not open" {
+      // handle
+    } else {
+      panic("unknown error")
+    }
+  }
+}
+```
+
+</td><td>
+
+```go
+// package foo
+
+var ErrCouldNotOpen = errors.New("could not open")
+
+func Open() error {
+  return ErrCouldNotOpen
+}
+
+// package bar
+
+if err := foo.Open(); err != nil {
+  if errors.Is(err, foo.ErrCouldNotOpen) {
+    // handle
+  } else {
+    panic("unknown error")
+  }
+}
+```
+
+</td></tr>
+</tbody></table>
+
+If you have an error that clients may need to detect, and you would like to add more information to it (e.g., it is not a static string), then you should use a custom type.
+
+<table>
+<thead><tr><th>Bad</th><th>Good</th></tr></thead>
+<tbody>
+<tr><td>
+
+```go
+func open(file string) error {
+  return fmt.Errorf("file %q not found", file)
+}
+
+func use() {
+  if err := open("testfile.txt"); err != nil {
+    if strings.Contains(err.Error(), "not found") {
+      // handle
+    } else {
+      panic("unknown error")
+    }
+  }
+}
+```
+
+</td><td>
+
+```go
+type errNotFound struct {
+  file string
+}
+
+func (e errNotFound) Error() string {
+  return fmt.Sprintf("file %q not found", e.file)
+}
+
+func open(file string) error {
+  return errNotFound{file: file}
+}
+
+func use() {
+  if err := open("testfile.txt"); err != nil {
+    if _, ok := err.(errNotFound); ok {
+      // handle
+    } else {
+      panic("unknown error")
+    }
+  }
+}
+```
+
+</td></tr>
+</tbody></table>
+
+Be careful with exporting custom error types directly since they become part of the public API of the package. It is preferable to expose matcher functions to check the error instead.
```

Comment:
Second sentence on it's own line.

Also, we pretty rarely use sentinel errors, but we do have a few, and we didn't use the matcher functions, so this is preaching something we don't do :D

Reply:
We do not do it currently because we don't use any custom error types. Once again, this is true for when we define custom error types and need to decide what to consider to match for `errors.Is` (whether just the error string matters or the attributes do as well.)

Reply:
We do however have them.. Again, not a lot, but there are a few cases..

models/dps/errors.go:
```go
// Sentinel errors.
var (
	ErrFinished    = errors.New("finished")
	ErrUnavailable = errors.New("unavailable")
)
```

Usage in service/mapper/transitions.go:
```go
header, err := t.chain.Header(s.height)
	if errors.Is(err, dps.ErrUnavailable) {
		log.Debug().Msg("waiting for next header")
		time.Sleep(t.cfg.WaitInterval)
		return nil
	}
	if err != nil {
		return fmt.Errorf("could not get header: %w", err)
	}
```

Or in flow-rosetta, rosetta/failure/invalid_block_hash.go:
```go
// InvalidBlockHash is the error for a block hash of invalid length.
type InvalidBlockHash struct {
	Description Description
	WantLength  int
	HaveLength  int
}

// Error implements the error interface.
func (i InvalidBlockHash) Error() string {
	return fmt.Sprintf("invalid block hash length (want: %d, have: %d): %s", i.WantLength, i.HaveLength, i.Description)
}
```

vs use in api/rosetta/errors.go:
```go
var ibErr failure.InvalidBlockHash
	if errors.As(err, &ibErr) {
		return echo.NewHTTPError(statusBadRequest, invalidFormat(ibErr.Description.Text,
			withDetail("want_length", ibErr.WantLength),
			withDetail("have_length", ibErr.HaveLength),
		))
	}
```

I do agree that we had no real need for it, nor I would expect someone to start using Flow DPS as a library, it just caught my eye that it's something we don't *technically* abide by.

Reply:
Ah it's true that we did not implement it. I thought we did tbh 😄 I feel like if we expect people to use this package, we should implement matcher functions to explicitly define whether or not an error can match another based on its attributes.